eth/client.go

```diff
@@ -79,13 +80,13 @@ func (c *PooledClient) inferError(err error) error {
 
 	switch {
 	case strings.Contains(err.Error(), "Cost of transaction exceeds sender balance"):
-		return concurrent.ErrCannotRecover{Cause: ErrExceedsBalance}
+		return errors.WithStack(concurrent.ErrCannotRecover{Cause: ErrExceedsBalance})
 	case strings.Contains(err.Error(), "Requested gas greater than block gas limit"):
-		return concurrent.ErrCannotRecover{Cause: ErrExceedsBlockLimit}
+		return errors.WithStack(concurrent.ErrCannotRecover{Cause: ErrExceedsBlockLimit})
 	case strings.Contains(err.Error(), "Invalid transaction nonce"):
-		return concurrent.ErrCannotRecover{Cause: ErrInvalidNonce}
+		return errors.WithStack(concurrent.ErrCannotRecover{Cause: ErrInvalidNonce})
```

Comment:
I think the whole point of this function is to wrap only certain errors into `concurrent.ErrCannotRecover` as a way of flagging them. `oasis-gateway/concurrent/conc.go:141` relies on this. Is `errors.WithStack` a new class? (sorry no time to check rn)
If not, would it work to make `concurrent.ErrCannotRecover` the outer error? (I haven't thought it through). Or maybe change `conc.go` to inspect the inside of `errors.WithStack`.

Reply:
Yep you're correct. I reverted this and added the stack in the calling function instead.